0000-plugins.md

```diff
@@ -0,0 +1,169 @@
+- Database schema changes across Lemmy versions, which would make it very difficult to make plugins compatible with new Lemmy versions, and prevent instances from upgrading
+- Plugins could accidentally delete data or break the database
+
+As such we need to use a different approach, by exposting specific database calls from Lemmy via [host functions](https://extism.org/docs/concepts/host-functions). For the beginning the following functions will be exposed:
+
+- Instance::read_or_create 
+
+More functions may be added if they are needed for the initial plugins.
+
+## Initial Plugins
+
+These can mainly serve as proof of concept, and as examples which can be modified by plugin developers. Include integration tests to ensure that functionality works with new Lemmy changes. These plugins should be written in a variety of languages, and live in a separate repo `lemmy-plugin-examples`.
+
+- Block only downvotes from a specific instance
+- Replace words in post or comment text (eg "cloud to butt")
+- Replace Youtube URLs in post links with [Invidious](https://invidious.io/)
+- Push notifications (send data of new posts to localhost url, [#2631](https://github.com/LemmyNet/lemmy/issues/2631))
+- Language detection (automatically apply language tag to posts/comments, [#2870](https://github.com/LemmyNet/lemmy/issues/2870))
+- Archive previous versions of posts when they are edited
+- Reject comments from specific user IDs (like [!santabot@slrpnk.net](https://slrpnk.net/post/11069853))
+- Additional suggestions welcome
+
+
+## Plugin Licensing
+
+Plugins can use any [OSI-approved open source licenses](https://opensource.org/licenses). Unlike Lemmy's AGPL license, most of them don't require sharing the source code with people who use it over a network, so the source code only needs to be made available to server admins. This also allows for for paid plugins.
+
+# Drawbacks
+
+Plugins may have a negative impact on Lemmy's performance. It is up to plugin developers and instance admins to ensure that plugins run adequately. In any case plugins are only called for POST actions but not for GET. This means slow plugins might result in longer waiting time to submit posts, but passive browsing will be unaffected.
+
+# Rationale and alternatives
+
+TODO
+
+# Prior art
+
+Plugins for other Fediverse platforms, all based on PHP:
+
+- https://github.com/friendica/friendica-addons
+- https://codeberg.org/hubzilla/hubzilla-addons
+- https://nextcloud.github.io/news/features/plugins/
+- https://developer.wordpress.org/plugins/
+
+# Unresolved questions
+
+What are the licensing requirements for plugins, if any? Are proprietary plugins allowed or do they need to be under AGPL like Lemmy? May have to ask Extism maintainers about legal conditions.
+
+# Future possibilities
+
```

Comment:
I suggest we define the concept of plugin metadata from the start. Each plugin could be annotated by one or more of the following fields, inspired by apt (see also: other package managers):

* an ID (using Java-like reverse DNS namespace notation)
* a short name that fits a list
* an icon
* a short, few line description of what it does, possibly including indexed keywords (hashtags) if that's not present in a designated field
* an easily parsable version number
* copyright declaration (license SPDX or link, authors, date)
* link to main landing page of own website
* link to source code
* link to a bug tracker, support contact channel or group, also how to report security issues
* an exhaustive list of all API hooks the plugins would like permission to access during its operation (both to allow the admin to review it before installation and as defense in depth against bugs)
* a list of plugins (possibly with minimum & maximum version number) which are required, suggested, conflicting and replaced by this one respectively

Reply:
Added a note about metadata near the beginning under File Structure. Though I want to keep it simple with only a few fields.